src/rust/scheduler/scheduler.rs

```diff
@@ -177,6 +175,15 @@ impl Scheduler {
         (page_index, page_offset)
     }
 
+    // Add new page(s) to hold this future's status if the current page is filled. This may result in addition of
+    // multiple pages because of the gap between the pin slab index and the current page index.
+    fn add_new_pages_to_accommodate_index_gap(&self, pin_slab_index: usize) {
+        let mut page_ref: RefMut<Vec<WakerPageRef>> = self.page_refs.borrow_mut();
+        while pin_slab_index >= page_ref.len() << WAKER_BIT_LENGTH_SHIFT {
+            page_ref.push(WakerPageRef::default());
+        }
+    }
+
```

Comment:
I don't see any reason why this couldn't be inline as it was before.

Reply:
I'm moving as many lower levels of abstractions as possible to their specific scope so that the semantics of the calling function stand out clearly. This is a lower level abstraction that is complex enough that it deserves a separate logical block (a function in this case).